### PR TITLE
Fix the machinery for building OCI images

### LINKED ISSUE #461: image_build 'artifacts' are not found during build time
Author: @dkliban (dkliban@redhat.com)

Redmine Issue: 6745, https://pulp.plan.io/issues/6745

---

```
$ http --auth admin:password --form :8080/pulp/api/v3/repositories/container/container/856132db-f1da-4b94-b6f2-0ee8196513ae/build_image/ containerfile@./Containerfile artifact="{\"/pulp/api/v3/artifacts/efb8d793-0bea-4d2d-9290-7846850cd802/\": \"foo/bar/example.txt\"}"

HTTP/1.1 202 Accepted
Allow: POST, OPTIONS
Connection: keep-alive
Content-Length: 67
Content-Type: application/json
Date: Thu, 14 May 2020 19:06:33 GMT
Server: nginx/1.18.0
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "task": "/pulp/api/v3/tasks/391134c2-46dc-4b01-b890-17d3d2ea9af5/"
}


```

```
$ http --auth admin:password  :8080/pulp/api/v3/tasks/391134c2-46dc-4b01-b890-17d3d2ea9af5/
HTTP/1.1 200 OK
Allow: GET, PATCH, DELETE, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 1687
Content-Type: application/json
Date: Thu, 14 May 2020 19:06:52 GMT
Server: nginx/1.18.0
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "child_tasks": [],
    "created_resources": [],
    "error": {
        "description": "b'Getting image source signatures\\nCopying blob sha256:e6a50b627bcb03d96996bb8e836ecb178eae7425636e3424d9e8d33a918768dd\\nCopying config sha256:8114e7c1868bd9b15188df8c34953b151b81f90a73f16bc00aaec227bd500cc9\\nWriting manifest to image destination\\nStoring signatures\\nerror building at STEP \"COPY foo/bar/example.txt /inside-image.txt\": no files found matching \"/var/lib/pulp/tmp/13105@f9c8c543863e/391134c2-46dc-4b01-b890-17d3d2ea9af5/foo/bar/example.txt\": no such file or directory\\nlevel=error msg=\"exit status 1\"\\n'",
        "traceback": "  File \"/usr/local/lib/python3.7/site-packages/rq/worker.py\", line 886, in perform_job\n    rv = job.perform()\n  File \"/usr/local/lib/python3.7/site-packages/rq/job.py\", line 664, in perform\n    self._result = self._execute()\n  File \"/usr/local/lib/python3.7/site-packages/rq/job.py\", line 670, in _execute\n    return self.func(*self.args, **self.kwargs)\n  File \"/usr/local/lib/python3.7/site-packages/pulp_container/app/tasks/builder.py\", line 124, in build_image_from_containerfile\n    raise Exception(bud_cp.stderr)\n"
    },
    "finished_at": "2020-05-14T19:06:44.994490Z",
    "name": "pulp_container.app.tasks.builder.build_image_from_containerfile",
    "parent_task": null,
    "progress_reports": [],
    "pulp_created": "2020-05-14T19:06:33.661619Z",
    "pulp_href": "/pulp/api/v3/tasks/391134c2-46dc-4b01-b890-17d3d2ea9af5/",
    "reserved_resources_record": [
        "/pulp/api/v3/repositories/container/container/856132db-f1da-4b94-b6f2-0ee8196513ae/"
    ],
    "started_at": "2020-05-14T19:06:33.822966Z",
    "state": "failed",
    "task_group": null,
    "worker": "/pulp/api/v3/workers/81de122d-2321-4c38-9f3a-9080f7cd9bba/"
}



```

### PR DESCRIPTION
**Pull request by lubosmj**
closes #461

### CODE REVIEW

**Comment by lubosmj** [comment]
I am waiting for the work on running functional tests within a pulp container to be completed to proceed.

File: pulp_container/tests/functional/api/test_build_images.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    monitor_task(build_response.task)
+
+    distribution = gen_object_with_cleanup(
+        container_distribution_api,
+        ContainerContainerDistribution(**gen_distribution(repository=repository.pulp_href)),
+    )
+
+    local_registry.pull(distribution.base_path)
+    image = local_registry.inspect(distribution.base_path)
+    assert image[0]["Config"]["Cmd"] == ["cat", "/tmp/inside-image.txt"]
```
Comment (ipanova) [inline pulp_container/tests/functional/api/test_build_images.py:63]: i think you might want to do `podman rmi distribution.base_path`

File: docs/installation.rst [hunk 2]
```diff
@@ -69,16 +69,9 @@ Run Services
...
-`buildah <https://github.com/containers/buildah/>`_ to build the container image. Buildah 1.14+
-must be installed on the same machine that is running pulpcore-worker processes.
-
-The pulpcore-worker processes needs to have `/usr/bin/` in its `PATH`. The user that is running
-pulpcore-worker process needs to be able to manage subordinate user ids and group ids. The range of
-subordinate user ids is specified in `/etc/subuid` and the range of subordinate group ids is
-specified in `/etc/subgid`. More details can be found in `buildah documentation <https://github.com
-/containers/libpod/blob/master/docs/tutorials/rootless_tutorial.md#enable-user-namespaces-on-rhel7-
-machines>`_.
+OCI Container Image building
```
Comment (ipanova) [inline docs/installation.rst:72]: you missed this file https://github.com/pulp/pulp_container/blob/be3995514e7d7fa3080db13a934a89cf8bbedf53/docs/workflows/build-containerfile.rst

File: pulp_container/app/tasks/builder.py [hunk 3]
```diff
@@ -76,20 +76,26 @@ def add_image_from_directory_to_repository(path, repository, tag):
...
+        config_blob = get_or_create_blob(manifest_json["config"], manifest, path)
+        manifest.config_blob = config_blob
+        manifest.save()
+        new_repo_version.add_content(Blob.objects.filter(pk=config_blob.pk))
+
+        blobs_to_add = []
+        for layer in manifest_json["layers"]:
+            blobs_to_add.append(get_or_create_blob(layer, manifest, path))
+
+        new_repo_version.add_content(Blob.objects.filter(pk__in=blobs_to_add))
```
Comment (ipanova) [inline pulp_container/app/tasks/builder.py:92]: you are still calling add_content here 4 times. collect all the content pks are i've suggested in previous comment

File: pulp_container/app/tasks/builder.py [hunk 4]
```diff
@@ -73,23 +73,27 @@ def add_image_from_directory_to_repository(path, repository, tag):
...
+        manifest.config_blob = config_blob
+        manifest.save()
+
+        pks_to_add = []
+        for layer in manifest_json["layers"]:
+            pks_to_add.append(get_or_create_blob(layer, manifest, path).pk)
+
+        pks_to_add.extend([manifest.pk, tag.pk, config_blob.pk])
+        new_repo_version.add_content(Content.objects.filter(pk__in=pks_to_add))
+
```
Comment (ipanova) [inline pulp_container/app/tasks/builder.py:91]: sweet!

**Review by ipanova — APPROVED** [review]

File: pulp_container/app/tasks/builder.py (hunk 4 above)
Comment (lubosmj) [reply pulp_container/app/tasks/builder.py:91]: 🥇

**Pull request merged**